sockeye/encoder.py

```diff
@@ -413,3 +470,173 @@ def get_rnn_cells(self) -> List[mx.rnn.BaseRNNCell]:
         Returns a list of RNNCells used by this encoder.
         """
         return self.forward_rnn.get_rnn_cells() + self.reverse_rnn.get_rnn_cells()
+
+
+class CharacterSequenceEncoder(Encoder):
+    """
+    An encoder that maps a sequence of character embeddings to a shorter
+    sequence of segment embeddings using convolutional, pooling, and highway
+    layers.
+        * "Fully Character-Level Neural Machine Translation without Explicit Segmentation"
+          Jason Lee; Kyunghyun Cho; Thomas Hofmann (https://arxiv.org/pdf/1610.03017.pdf)
+
+    :param num_embed: Input embedding size.
+    :param max_filter_width: Maximum filter width for convolutions.
+    :param num_filters: Number of filters of each width.
+    :param pool_stride: Stride for pooling layer after convolutions.
+    :param num_highway_layers: Number of highway layers for segment embeddings.
+    :param prefix: Name prefix for symbols of this encoder.
+    :param dropout: Dropout probability.
+    """
```

Comment:
Right now this is a straightforward implementation of Figure 1 in the paper.  There's some room for optimization if people start using this regularly.